crates/core/liveview-native-core-swift/Sources/LiveViewNativeCore/Support.swift

```diff
@@ -0,0 +1,147 @@
+import Foundation
+
+class SimpleHandler: DocumentChangeHandler {
+    var callback: (Document, NodeRef) -> ()
+    init (
+        _ callback: @escaping (Document, NodeRef) -> ()
+    ) {
+       self.callback = callback
+    }
+    func handle(_ context: Document, _ changeType: ChangeType, _ node: NodeRef, _ parent: NodeRef?) {
+        switch changeType {
+        case .add:
+            self.callback(context, parent!)
+        case .remove:
+            self.callback(context, parent!)
+        case .change:
+            self.callback(context, node)
+        case .replace:
+            self.callback(context, parent!)
+        }
+    }
+
+}
+
+extension Document {
+    public subscript(ref: NodeRef) -> Node {
+        let data = self.get(ref)
+        return Node(self, ref, data)
+    }
+    public static func parseFragmentJson(payload: [String: Any]) throws -> Document {
+        let jsonData = try JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted)
+        let payload = String(data: jsonData, encoding: .utf8)!
+        return try Document.parseFragmentJson(payload)
+    }
+    public func mergeFragmentJson(
+        _ payload: [String: Any]
+        ) throws {
+        let jsonData = try JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted)
+        let payload = String(data: jsonData, encoding: .utf8)!
+
+        return try self.mergeFragmentJson(payload)
+    }
+
+    public func on(_ event: EventType, _ callback: @escaping (Document, NodeRef) -> ()) {
```

Comment:
It'd be nice if we could have an `AsyncStream` for this API.

Reply:
Well I refactored this a few times to fix a deadlock. When `mergeFragmentJson` is called, it's locking a mutex for the Document on the rust side as it's mutating the document. Today, I pulled in latest main https://github.com/liveview-native/liveview-client-swiftui/pull/1287 and found that [recent changes to `ObservedElement`](https://github.com/liveview-native/liveview-client-swiftui/pull/1300/files#diff-f8986450f426e84938cdf7afe8bef21ba6f2d3f7bf75e3ba66ddc6baf8a0212eR117-R127) caused the deadlock. This is in part because when the closure is called on an incremental edit update, [`elementChanged`](https://github.com/liveview-native/liveview-client-swiftui/blob/35dfb555b241d8fb7bc9dd2bf4e614681e3b3b6a/Sources/LiveViewNative/Coordinators/LiveViewCoordinator.swift#L42-L49) which triggers [the sink closure on the `ObservedElement`](https://github.com/liveview-native/liveview-client-swiftui/blob/35dfb555b241d8fb7bc9dd2bf4e614681e3b3b6a/Sources/LiveViewNative/Property%20Wrappers/ObservedElement.swift#L125-L126) which uses Document.subscript and Document.children which require read locks on the same mutex that's currently locked for editing. I wonder if an `AsyncStream` for this API would help?

Reply:
`elementChanged` could be refactored to publish the element, instead of just `Void`. Then the child caching could be done on read instead of when the change occurs.